plugins/spi/SPIOutput.cpp

```diff
@@ -180,26 +185,67 @@ SPIOutput::SPIOutput(const UID &uid, SPIBackendInterface *backend,
-  personalities.push_back(Personality(m_pixel_count * LPD8806_SLOTS_PER_PIXEL,
-                                      "LPD8806 Individual Control"));
-  personalities.push_back(Personality(LPD8806_SLOTS_PER_PIXEL,
-                                      "LPD8806 Combined Control"));
-  personalities.push_back(Personality(m_pixel_count * P9813_SLOTS_PER_PIXEL,
-                                      "P9813 Individual Control"));
-  personalities.push_back(Personality(P9813_SLOTS_PER_PIXEL,
-                                      "P9813 Combined Control"));
-  personalities.push_back(Personality(m_pixel_count * APA102_SLOTS_PER_PIXEL,
-                                      "APA102 Individual Control"));
-  personalities.push_back(Personality(APA102_SLOTS_PER_PIXEL,
-                                      "APA102 Combined Control"));
+  // personality description is max 32 characters
+
+  personalities.insert(personalities.begin() + PERS_WS2801_INDIVIDUAL - 1,
+                      Personality(m_pixel_count * WS2801_SLOTS_PER_PIXEL,
+                                  "WS2801 Individual Control"));
+  personalities.insert(personalities.begin() + PERS_WS2801_COMBINED - 1,
+                      Personality(WS2801_SLOTS_PER_PIXEL,
+                                  "WS2801 Combined Control"));
+
+  personalities.insert(personalities.begin() + PERS_LDP8806_INDIVIDUAL - 1,
+                      Personality(m_pixel_count * LPD8806_SLOTS_PER_PIXEL,
+                                  "LPD8806 Individual Control"));
+  personalities.insert(personalities.begin() + PERS_LDP8806_COMBINED - 1,
+                      Personality(LPD8806_SLOTS_PER_PIXEL,
+                                  "LPD8806 Combined Control"));
+
+  personalities.insert(personalities.begin() + PERS_P9813_INDIVIDUAL - 1,
+                      Personality(m_pixel_count * P9813_SLOTS_PER_PIXEL,
+                                  "P9813 Individual Control"));
+  personalities.insert(personalities.begin() + PERS_P9813_COMBINED - 1,
+                      Personality(P9813_SLOTS_PER_PIXEL,
+                                  "P9813 Combined Control"));
+
+  personalities.insert(personalities.begin() + PERS_APA102_INDIVIDUAL - 1,
+                      Personality(m_pixel_count * APA102_SLOTS_PER_PIXEL,
+                                  "APA102 Individual Control"));
+
+  ola::rdm::SlotDataCollection::SlotDataList sd_APA102_COMBINED;
+  sd_APA102_COMBINED.push_back(
+      ola::rdm::SlotData::PrimarySlot(ola::rdm::SD_COLOR_ADD_RED, 0));
+  sd_APA102_COMBINED.push_back(
+      ola::rdm::SlotData::PrimarySlot(ola::rdm::SD_COLOR_ADD_GREEN, 0));
+  sd_APA102_COMBINED.push_back(
+      ola::rdm::SlotData::PrimarySlot(ola::rdm::SD_COLOR_ADD_BLUE, 0));
+  personalities.insert(personalities.begin() + PERS_APA102_COMBINED - 1,
+                      Personality(APA102_SLOTS_PER_PIXEL,
+                                  "APA102 Combined Control",
+      ola::rdm::SlotDataCollection(sd_APA102_COMBINED)));
```

Comment:
The line ola::rdm::SlotDataCollection(sd_APA102_COMBINED) can become sd_APA102_COMBINED.

Reply:
Done

Reply:
hm - i was to fast with the commit - 
for me this results in an error:
```
plugins/spi/SPIOutput.cpp: In constructor ‘ola::plugin::spi::SPIOutput::SPIOutput(const ola::rdm::UID&, ola::plugin::spi::SPIBackendInterface*, const ola::plugin::spi::SPIOutput::Options&)’:
plugins/spi/SPIOutput.cpp:225:53: error: no matching function for call to ‘ola::rdm::Personality::Personality(const uint16_t&, const char [24], ola::rdm::SlotDataCollection::SlotDataList&)’
                                   sd_APA102_COMBINED));
                                                     ^
In file included from ./include/ola/rdm/ResponderHelper.h:35:0,
                 from plugins/spi/SPIOutput.cpp:42:
./include/ola/rdm/ResponderPersonality.h:41:5: note: candidate: ola::rdm::Personality::Personality(uint16_t, const string&, const ola::rdm::SlotDataCollection&)
     Personality(uint16_t footprint, const std::string &description,
     ^~~~~~~~~~~
./include/ola/rdm/ResponderPersonality.h:41:5: note:   no known conversion for argument 3 from ‘ola::rdm::SlotDataCollection::SlotDataList {aka std::vector<ola::rdm::SlotData>}’ to ‘const ola::rdm::SlotDataCollection&’
./include/ola/rdm/ResponderPersonality.h:40:5: note: candidate: ola::rdm::Personality::Personality(uint16_t, const string&)
     Personality(uint16_t footprint, const std::string &description);
     ^~~~~~~~~~~
./include/ola/rdm/ResponderPersonality.h:40:5: note:   candidate expects 2 arguments, 3 provided
./include/ola/rdm/ResponderPersonality.h:38:7: note: candidate: ola::rdm::Personality::Personality(const ola::rdm::Personality&)
 class Personality {
       ^~~~~~~~~~~
./include/ola/rdm/ResponderPersonality.h:38:7: note:   candidate expects 1 argument, 3 provided
./include/ola/rdm/ResponderPersonality.h:38:7: note: candidate: ola::rdm::Personality::Personality(ola::rdm::Personality&&)
./include/ola/rdm/ResponderPersonality.h:38:7: note:   candidate expects 1 argument, 3 provided
plugins/spi/SPIOutput.cpp:243:57: error: no matching function for call to ‘ola::rdm::Personality::Personality(const uint16_t&, const char [33], ola::rdm::SlotDataCollection::SlotDataList&)’
                                    sd_APA102_PB_COMBINED));
                                                         ^
In file included from ./include/ola/rdm/ResponderHelper.h:35:0,
                 from plugins/spi/SPIOutput.cpp:42:
./include/ola/rdm/ResponderPersonality.h:41:5: note: candidate: ola::rdm::Personality::Personality(uint16_t, const string&, const ola::rdm::SlotDataCollection&)
     Personality(uint16_t footprint, const std::string &description,
     ^~~~~~~~~~~
./include/ola/rdm/ResponderPersonality.h:41:5: note:   no known conversion for argument 3 from ‘ola::rdm::SlotDataCollection::SlotDataList {aka std::vector<ola::rdm::SlotData>}’ to ‘const ola::rdm::SlotDataCollection&’
./include/ola/rdm/ResponderPersonality.h:40:5: note: candidate: ola::rdm::Personality::Personality(uint16_t, const string&)
     Personality(uint16_t footprint, const std::string &description);
     ^~~~~~~~~~~
./include/ola/rdm/ResponderPersonality.h:40:5: note:   candidate expects 2 arguments, 3 provided
./include/ola/rdm/ResponderPersonality.h:38:7: note: candidate: ola::rdm::Personality::Personality(const ola::rdm::Personality&)
 class Personality {
       ^~~~~~~~~~~
./include/ola/rdm/ResponderPersonality.h:38:7: note:   candidate expects 1 argument, 3 provided
./include/ola/rdm/ResponderPersonality.h:38:7: note: candidate: ola::rdm::Personality::Personality(ola::rdm::Personality&&)
./include/ola/rdm/ResponderPersonality.h:38:7: note:   candidate expects 1 argument, 3 provided
```

Reply:
Oops, my apologies, sorry about that. Perhaps we should have a Personality constructor that takes a vector directly?

Reply:
seems like a good idea ;-)
i can try to look at this next week.
would it be enough to add a additional overloaded constructor and from this one use the cast and call the current one? (i have not looked at the code yet..)

Reply:
Yes I think so, but don't trust what I said given the earlier comment! :)